0002-RCBS-Crypto-building-block.md

```diff
@@ -0,0 +1,675 @@
+
+> In the future, we can support other authenticated ciphers such as AES-CBC with HMAC-SHA256.
+
+#### MAC
+
+The third and final line is the MAC for the header, which is computed with HMAC-SHA-256 over the previous 2 lines (including the final newline character) with a key that is derived from the (plain-text) File Key with HKDF-SHA-256:
+
+```text
+mac-key = HKDF-SHA-256(ikm = file key, salt = empty, info = "header")
+MAC = HMAC-SHA-256(key = mac-key, message = first 2 lines of the header)
+```
+
+> HKDF-SHA-256 is a key derivation function based on HMAC with SHA-256. See [RFC 5869 ("HMAC-based Extract-and-Expand Key Derivation Function (HKDF)")](https://www.rfc-editor.org/rfc/rfc5869.html). Being based on HMAC, it's not vulnerable to length-extension attacks, so we do not consider necessary using SHA-512 and truncating the output to 256-bits.
+
+Note that there's one newline character (0x0A) at the end of the MAC, which concludes the header.
+
+> Because each JSON encoder could produce a slightly different output, when verifying the manifest the MAC should be computed on the exact manifest string as included in the header. Verifiers should not re-encode the raw message as JSON.
+
+### Binary payload
+
+The binary payload begins immediately after the header (after the 3rd newline character) and it includes the each segment of data encrypted:
+
+```text
+segment_0 || segment_1 || ... || segment_k
+```
+
+### Segments
+
+The plaintext is chunked into segments 64KB (65,536 bytes) each; the last segment may be shorter. Segments must never be empty, unless the entire file is empty.
+
+> Because segments are 64KB each, and we can have up to 2^32 segments, the maximum size of the encrypted message is 256TB.
+
+Each segment of plaintext is encrypted independently and stored together with its authentication tag:
+
+```text
+encrypted chunk || tag
+```
+
+Segments are encrypted with a **Payload Key (PK)** that is derived from the (plain-text) File Key and the nonce prefix:
+
+```text
+payload-key = HKDF-SHA-256(ikm = file key, salt = nonce prefix, info = "payload")
+```
+
+Each segment is encrypted using a different 12-byte nonce / initialization vector:
+
+```text
+nonce_prefix || i || last_segment
+```
+
+Where:
+
+- `nonce_prefix` (7 bytes) is the nonce prefix from the header
+- `i` (4 bytes) is the sequence number, as a 32-bit unsigned integer counter, encoded as big-endian. The first segment has sequence number 0, and it increases.
+- `last_segment` (1 byte) is `0x01` if this is the last segment, or `0x00` otherwise
+
+## Components
+
+Components in dapr/components-contrib implement low-level primitives only, while all higher-level operations are performed by the runtime, so they are executed in a consistent way across all backends/services. This is because the job of the components is limited to actually interacting with the key vaults, and everything else is best handled by the runtime.
```

Comment:
Agree 100%. So, there will be a `crypto` type for components and 2 APIs will consume that (subtle and high-level), correct?

Reply:
Correct. The diagram shows the flow: https://github.com/dapr/proposals/blob/c8f398716815ab51b67478cadde579c8384c3e45/resources/0002-RCBS-Crypto-building-block/data-flow.png